worlds/ufo50/games/vainger/rules.py

```diff
@@ -0,0 +1,248 @@
+                    rule = lambda state: state.count(security_clearance, player) >= 1) 
+    verdeswarea.connect(verdei9,
+                        rule = lambda state: state.count(security_clearance, player) >= 2 
+                                             and state.has("Vainger - VerdeI9 - Sura Defeated", player)) # genepod only exists after boss kill
+    # the spike tank strat is unreasonable coming from the left, so the fact that a player coming from the left *might* have used hot-shot
+    # to get here is irrelevant.
+    verdeswarea.connect(verdeh7loc,
+                        rule = lambda state: state.has(force_mod, player)) # spike-ng.
+    verdei7.connect(verdeh7loc,
+                    rule = lambda state: state.has(force_mod, player) or spike_tank(state, world)) # here spike tanking is reasonable
+
+    #TODO: check how hard this hell run is
+    latomc9.connect(latomf7,
+                    rule = lambda state: state.has(heat_mod, player) or hell_run(10, False, state, world)) # hot-shot, magmatek, or hell run
+    latomc9.connect(latomf5,
+                    rule = lambda state: state.has_all([heat_mod, pulse_mod], player)) # hot-shot and thunder
+    latomf7.connect(latomc6,
+                    rule = lambda state: state.count(security_clearance, player) >= 3
+                                         and state.has_any([pulse_mod, multi_mod], player)) # thunder or tri-shot
+    latomf7.connect(latomd3,
+                    rule = lambda state: state.has_any([pulse_mod, multi_mod], player)) # thunder or tri-shot
+    latomf7.connect(latomf5,
+                    rule = lambda state: state.has(pulse_mod, player)) # thunder or possibly tanking an electrical arc later
+    latomd3.connect(latomf5)
+    latomd3.connect(latomc6,
+                    rule = lambda state: state.count(security_clearance, player) >= 3)
+    latomf5.connect(latomd3)
+    latomf5.connect(latomc6,
+                    rule = lambda state: state.count(security_clearance, player) >= 2 and state.has(heat_mod, player)) # hot-shot
+    latomc6.connect(latomd5,
+                    rule = lambda state: state.count(security_clearance, player) >= 3 and state.has("Vainger - LatosD5 - Boss Defeated", player))
+    # TODO: the normal route is one-way; does the miniboss block you from doing the upper route in reverse?
+    latomf5.connect(latomi4,
+                    rule = lambda state: state.has(pulse_mod, player)) # NOTE: thunder required to prevent a softlock; this means vanilla pulse mod will be impossible.
+    latomc6.connect(latomd6area,
+                    rule = lambda state: state.has(heat_mod, player)) # hot-shot for the shortcut C6 -> D6
+    latomf5.connect(latomd6area,
+                    rule = lambda state: state.count(security_clearance, player) >= 2)
+    
+    def sr(loc: str, rule: CollectionRule = lambda state: True):
+        set_rule(world.get_location(f"Vainger - {loc}"), rule)
+
+    # LatomD3
+    sr("LatomA4 - Shield Upgrade")
```

Comment:
You don't need to set the rule to True -- it's like that by default. Would recommend either deleting these or commenting them out (I could see wanting them for the explicitness)

Reply:
My own taste would be for it to be a compile error not to set all of them, but when in Python,

I'll comment them out.